src/coreclr/jit/importer_vectorization.cpp

```diff
@@ -0,0 +1,741 @@
+//    type -       type for the IND node
+//    offset -     offset for the data pointer
+//    value -      constant value to compare against
+//
+// Return Value:
+//    A tree with indirect load and comparison
+//
+static GenTree* impCreateCompareInd(Compiler* comp, GenTree* obj, var_types type, ssize_t offset, ssize_t value)
+{
+    GenTree* offsetTree    = comp->gtNewIconNode(offset, TYP_I_IMPL);
+    GenTree* addOffsetTree = comp->gtNewOperNode(GT_ADD, TYP_BYREF, obj, offsetTree);
+    GenTree* indirTree     = comp->gtNewIndir(type, addOffsetTree);
+    GenTree* valueTree     = comp->gtNewIconNode(value, type);
+    if (varTypeIsSmall(indirTree))
+    {
+        indirTree = comp->gtNewCastNode(TYP_INT, indirTree, true, TYP_UINT);
+        valueTree->ChangeType(TYP_INT);
+    }
+    return comp->gtNewOperNode(GT_EQ, TYP_INT, indirTree, valueTree);
+}
+
+//------------------------------------------------------------------------
+// impExpandHalfConstEqualsSWAR: Attempts to unroll and vectorize
+//    Equals against a constant WCHAR data for Length in [1..8] range
+//    using SWAR (a sort of SIMD but for GPR registers and instructions)
+//
+// Arguments:
+//    data -       Pointer to a data to vectorize
+//    cns -        Constant data (array of 2-byte chars)
+//    len -        Number of chars in the cns
+//    dataOffset - Offset for data
+//
+// Return Value:
+//    A pointer to the newly created SWAR node or nullptr if unrolling is not
+//    possible or not profitable
+//
+// Notes:
+//    This function doesn't check obj for null or its Length, it's just an internal helper
+//    for impExpandHalfConstEquals
+//
+GenTree* Compiler::impExpandHalfConstEqualsSWAR(GenTree* data, WCHAR* cns, int len, int dataOffset)
+{
+    assert(len >= 1 && len <= 8);
+
+// Compose Int32 or Int64 values from ushort components
+#define MAKEINT32(c1, c2) ((UINT64)c2 << 16) | ((UINT64)c1 << 0)
+#define MAKEINT64(c1, c2, c3, c4) ((UINT64)c4 << 48) | ((UINT64)c3 << 32) | ((UINT64)c2 << 16) | ((UINT64)c1 << 0)
+
+    if (len == 1)
+    {
+        //   [ ch1 ]
+        //   [value]
+        //
+        return impCreateCompareInd(this, data, TYP_SHORT, dataOffset, cns[0]);
+    }
+    if (len == 2)
+    {
+        //   [ ch1 ][ ch2 ]
+        //   [   value    ]
+        //
+        const UINT32 value = MAKEINT32(cns[0], cns[1]);
+        return impCreateCompareInd(this, data, TYP_INT, dataOffset, value);
+    }
+#ifdef TARGET_64BIT
+    if (len == 3)
+    {
+        // handle len = 3 via two Int32 with overlapping:
+        //
+        //   [ ch1 ][ ch2 ][ ch3 ]
+        //   [   value1   ]
+        //          [   value2   ]
+        //
+        UINT32   value1      = MAKEINT32(cns[0], cns[1]);
+        UINT32   value2      = MAKEINT32(cns[1], cns[2]);
+        GenTree* firstIndir  = impCreateCompareInd(this, data, TYP_INT, dataOffset, value1);
+        GenTree* secondIndir = impCreateCompareInd(this, gtClone(data), TYP_INT, dataOffset + 2, value2);
```

Comment:
Ditto on `data` and `gtClone`.

`+ 2` here because each char is 2 bytes? Might make that clearer somehow.

Is this better than comparing 2, then 1?

Reply:
> + 2 here because each char is 2 bytes? Might make that clearer somehow.

There is a visualization on top of that expression:
```
        // handle len = 3 via two Int32 with overlapping:
        //
        //   [ ch1 ][ ch2 ][ ch3 ]
        //   [   value1   ]
        //          [   value2   ]
        //
```
I'll add an additional comment.

> Is this better than comparing 2, then 1?

Yes, INT+SHORT ends up with a redundant `movzx` in comparison with INT+INT